graphbolt/src/cuda/common.h

```diff
@@ -90,6 +91,44 @@ inline bool is_zero<dim3>(dim3 size) {
     }                                                                 \
   }
 
+/**
+ * @brief This class is designed to handle the copy operation of a single
+ * scalar_t item from a given CUDA device pointer. Later, if the object is cast
+ * into scalar_t, the value can be read.
+ *
+ * auto num_edges = cuda::CopyScalar(indptr.data_ptr<scalar_t>() +
+ *     indptr.size(0) - 1);
+ * // Perform many operations here, they will run as normal.
+ * // We finally need to read num_edges.
+ * auto indices = torch::empty(static_cast<scalar_t>(num_edges));
+ */
+template <typename scalar_t>
+struct CopyScalar {
+  CopyScalar(const scalar_t* device_ptr) : is_ready(false) {
+    pinned_scalar = torch::empty(
+        sizeof(scalar_t),
+        c10::TensorOptions().dtype(torch::kBool).pinned_memory(true));
+    auto stream = GetCurrentStream();
+    CUDA_CALL(cudaMemcpyAsync(
+        reinterpret_cast<scalar_t*>(pinned_scalar.data_ptr()), device_ptr,
+        sizeof(scalar_t), cudaMemcpyDeviceToHost, stream));
+    copy_event.record(stream);
+  }
+
+  operator scalar_t() {
+    if (!is_ready) {
+      copy_event.synchronize();
+      is_ready = true;
+    }
+    return reinterpret_cast<scalar_t*>(pinned_scalar.data_ptr())[0];
+  }
+
+ private:
+  torch::Tensor pinned_scalar;
+  at::cuda::CUDAEvent copy_event;
+  bool is_ready;
```

Comment:
nit: all internal variable should end with _, and rename is_ready -> synchronized_.
```suggestion
  torch::Tensor pinned_scalar_;
  at::cuda::CUDAEvent copy_event_;
  bool synchronized_;
```